qualtran/bloqs/block_encoding/product.py

```diff
@@ -0,0 +1,264 @@
+
+    @property
+    def signal_state(self) -> PrepareOracle:
+        """This method will be implemented in the future after PrepareOracle is updated for the BlockEncoding interface."""
+        raise NotImplementedError
+
+    def build_composite_bloq(
+        self, bb: BloqBuilder, system: SoquetT, **soqs: SoquetT
+    ) -> Dict[str, SoquetT]:
+        if (
+            is_symbolic(self.system_bitsize)
+            or is_symbolic(self.ancilla_bitsize)
+            or is_symbolic(self.resource_bitsize)
+        ):
+            raise DecomposeTypeError(f"Cannot decompose symbolic {self=}")
+        assert (
+            isinstance(self.system_bitsize, int)
+            and isinstance(self.ancilla_bitsize, int)
+            and isinstance(self.resource_bitsize, int)
+        )
+
+        n = len(self.block_encodings)
+        res_bits_used = 0
+        for i, u in enumerate(reversed(self.block_encodings)):
+            u_soqs = {"system": system}
+
+            # split ancilla register if necessary
+            if self.ancilla_bitsize > 0:
+                anc_bits = cast(int, u.ancilla_bitsize)
+                flag_bits = Register("flag_bits", dtype=QBit(), shape=(n - 1,))  # type: ignore
+                anc_used = Register("anc_used", dtype=QAny(anc_bits))
+                anc_unused_bits = self.ancilla_bitsize - (n - 1) - anc_bits
+                anc_unused = Register("anc_unused", dtype=QAny(anc_unused_bits))
+                anc_regs = [flag_bits]
+                if anc_bits > 0:
+                    anc_regs.append(anc_used)
+                if anc_unused_bits > 0:
+                    anc_regs.append(anc_unused)
+                anc_part = Partition(self.ancilla_bitsize, tuple(anc_regs))
+                anc_part_soqs = bb.add_d(anc_part, x=soqs["ancilla"])
+                if anc_bits > 0:
+                    u_soqs["ancilla"] = anc_part_soqs["anc_used"]
+
+            # split resource register if necessary
+            res_bits = cast(int, u.resource_bitsize)
+            if res_bits > 0:
+                res_before = Register("res_before", dtype=QAny(res_bits_used))
+                res = Register("res", dtype=QAny(res_bits))
+                res_bits_left = self.resource_bitsize - res_bits_used - res_bits
+                res_after = Register("res_after", dtype=QAny(res_bits_left))
+                res_regs = []
+                if res_bits_used > 0:
+                    res_regs.append(res_before)
+                res_regs.append(res)
+                res_bits_used += res_bits
+                if res_bits_left > 0:
+                    res_regs.append(res_after)
+                res_part = Partition(self.resource_bitsize, tuple(res_regs))
+                res_part_soqs = bb.add_d(res_part, x=soqs["resource"])
+                u_soqs["resource"] = res_part_soqs["res"]
+
+            # connect the constituent bloq
+            u_out_soqs = bb.add_d(u, **u_soqs)
+            system = u_out_soqs["system"]
```

Comment:
Am  I understanding this correct that you are using the simpler construction where each of the block encoding has it's own ancilla register and they don't share any ancillas? i.e. the construction from Figure 1 instead of Figure 2 ?

<img width="840" alt="image" src="https://github.com/quantumlib/Qualtran/assets/7863287/748ab7ed-09c4-42e0-93d6-04e29a19fa3c">



If that is the case, can we do the partition / unpartition of the ancilla registers once outside the loop instead of doing it as part of the loop? That would clean up the decomposition quite a bit I believe.

Reply:
nvm, seems like you are using decomposition from Fig 2. I missed the part where you were adding MCX gates below. 

In this case, I would do a parition of ancilla register into (a) Flag register and (b) Ancilla for individual bloq encodings outside the forloop and then add `n` bloqs for the `i` block encodings, each of them wrapped in an `AutoPartition`. The `MultiControlX` gate can use the unpartitioned ancilla directly. This would give us circuit diagrams very similar to the one drawn in the picture above.

Reply:
This may require some change to `AutoPartition`, and so I will do it in a follow-up PR.